### PR TITLE
Render Command for PAO

### PR DESCRIPTION
**Pull request by Tal-or**
Render command gets performance-profile manifest as an input and based on the provided profile,
it creates manifests yaml of the corresponding CR's which should have been created by the profile.

Signed-off-by: Talor Itzhak <titzhak@redhat.com>

### CODE REVIEW

File: cmd/render/render.go [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+
+type manifestable map[string]interface{}
+
+//NewRenderCommand creates a render command.
+func NewRenderCommand() *cobra.Command {
+	renderOpts := renderOpts{}
+
+	cmd := &cobra.Command{
+		Use:   "render",
+		Short: "Render performance-addon-operator CR manifests",
```
Comment (yanirq) [inline cmd/render/render.go:48]: "Render performance-addon-operator CR manifests"   Maybe drop "CR"  its a bit confusing, manifests should be enough. If you want to elaborate you can write exactly what types of manifests are created

File: cmd/render/render.go [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+	}
+
+	renderOpts.AddFlags(cmd.Flags())
+
+	return cmd
+}
+
+func (r *renderOpts) AddFlags(fs *pflag.FlagSet) {
+	fs.StringVar(&r.ppInputFilePath, "performance-profile-file-path", r.ppInputFilePath, "Input path for the Performance Profile manifest.")
+	fs.StringVar(&r.assetsDir, "assets-dir-path", components.AssetsDir, "Input path for the assets directory.")
```
Comment (yanirq) [inline cmd/render/render.go:68]: assets  or generated assets ?

File: cmd/render/render.go [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+	"github.com/ghodss/yaml"
+	performancev2 "github.com/openshift-kni/performance-addon-operators/api/v2"
+	"github.com/openshift-kni/performance-addon-operators/pkg/controller/performanceprofile/components"
+	"github.com/openshift-kni/performance-addon-operators/pkg/controller/performanceprofile/components/kubeletconfig"
+	"github.com/openshift-kni/performance-addon-operators/pkg/controller/performanceprofile/components/machineconfig"
+	"github.com/openshift-kni/performance-addon-operators/pkg/controller/performanceprofile/components/runtimeclass"
+	"github.com/openshift-kni/performance-addon-operators/pkg/controller/performanceprofile/components/tuned"
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+	"k8s.io/klog"
```
Comment (yanirq) [inline cmd/render/render.go:31]: I'm starting to see a lot  log references in different files,  not use its too much of a difference but maybe we should stay consistent  (log "github.com/sirupsen/logrus" for example is some references)   
@cynepco3hahue  what's your opinion here ?

File: cmd/render/render.go [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+}
+
+func (r *renderOpts) AddFlags(fs *pflag.FlagSet) {
+	fs.StringVar(&r.ppInputFilePath, "performance-profile-file-path", r.ppInputFilePath, "Input path for the Performance Profile manifest.")
+	fs.StringVar(&r.assetsDir, "assets-dir-path", components.AssetsDir, "Input path for the assets directory.")
+	fs.StringVar(&r.outputDir, "output-dir", r.outputDir, "Output path for the rendered manifests.")
+}
+
+func (r *renderOpts) Validate() error {
+	if len(r.ppInputFilePath) == 0 || len(r.outputDir) == 0 {
```
Comment (yanirq) [inline cmd/render/render.go:73]: why not split the validation here ? in case one of them is specified ?  maybe its more user friendly

File: cmd/render/render.go [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+*/
+
+package render
+
+import (
+	"fmt"
+	"io/ioutil"
+
+	"github.com/ghodss/yaml"
+	performancev2 "github.com/openshift-kni/performance-addon-operators/api/v2"
```
Comment (yanirq) [inline cmd/render/render.go:23]: maybe not for now, but we need to think what happens when we bump the performance profile version (if we add new apis)

**Comment by yanirq** [comment]
before jumping to e2e tests you can also provide unittests that checks generate examples  (compare using strings or regular expressions)

File: cmd/render/render.go (hunk 2 above)
Comment (Tal-or) [reply cmd/render/render.go:68]: This [asset](https://github.com/openshift-kni/performance-addon-operators/tree/master/build/assets) directory.
I'm not sure myself, but it's necessary for the components creation.

Comment (fromanirh) [reply cmd/render/render.go:68]: this should be renamed as `asset-input-dir`

File: cmd/render/render.go [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+func NewRenderCommand() *cobra.Command {
+	renderOpts := renderOpts{}
+
+	cmd := &cobra.Command{
+		Use:   "render",
+		Short: "Render performance-addon-operator CR manifests",
+		Run: func(cmd *cobra.Command, args []string) {
+
+			if err := renderOpts.Validate(); err != nil {
+				klog.Error(err)
```
Comment (fromanirh) [inline cmd/render/render.go:52]: why not Fatal here?

File: cmd/render/render.go [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+		},
+	}
+
+	renderOpts.AddFlags(cmd.Flags())
+
+	return cmd
+}
+
+func (r *renderOpts) AddFlags(fs *pflag.FlagSet) {
+	fs.StringVar(&r.ppInputFilePath, "performance-profile-file-path", r.ppInputFilePath, "Input path for the Performance Profile manifest.")
```
Comment (fromanirh) [inline cmd/render/render.go:67]: should this be part of `asset-input-dir` (e.g. read the file from there, fail if it is not inside)?

File: cmd/render/render.go [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+	fs.StringVar(&r.assetsDir, "assets-dir-path", components.AssetsDir, "Input path for the assets directory.")
+	fs.StringVar(&r.outputDir, "output-dir", r.outputDir, "Output path for the rendered manifests.")
```
Comment (fromanirh) [inline cmd/render/render.go:69]: this should be renamed as `asset-output-dir`

File: cmd/render/render.go (hunk 4 above)
Comment (fromanirh) [reply cmd/render/render.go:73]: +1 validate each field and return a separate error for each mandatory but missing field

File: cmd/render/render.go [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+	performanceTuned, err := tuned.NewNodePerformance(r.assetsDir, profile)
+	if err != nil {
+		return err
+	}
+	components[performanceTuned.Kind] = performanceTuned
+
+	runtimeClass := runtimeclass.New(profile, machineconfig.HighPerformanceRuntime)
+	components[runtimeClass.Kind] = runtimeClass
+
+	for name, component := range components {
```
Comment (fromanirh) [inline cmd/render/render.go:114]: not sure the kind is a great fit for the name

File: cmd/render/render.go [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+}
+
+func (r *renderOpts) Validate() error {
+	if len(r.ppInputFilePath) == 0 || len(r.outputDir) == 0 {
+		return fmt.Errorf("performance-profile-file-path and output-dir must be specified")
+	}
+	return nil
+}
+
+func (r *renderOpts) Run() error {
```
Comment (fromanirh) [inline cmd/render/render.go:79]: the problem here is we have a fully separate path wrt the online one. Say tomorrow we manage a new object, we will need to remember to add it here. How we can more likely we remember to do this?

File: main.go [hunk 11]
```diff
@@ -83,11 +85,33 @@ func printVersion() {
 }
 
 func main() {
+	// Add klog flags
+	klog.InitFlags(nil)
+
+	command := newPAOCommand()
```
Comment (fromanirh) [inline main.go:91]: I'd call this `newRootCommand` for consistency with `cobra` usage patterns

File: cmd/render/render.go (hunk 8 above)
Comment (Tal-or) [reply cmd/render/render.go:69]: The manifests also known as assets?

File: cmd/render/render.go (hunk 6 above)
Comment (Tal-or) [reply cmd/render/render.go:52]: Changed that on purpose, but I misunderstood the way it works.
Will replace back to Fatal

File: cmd/render/render.go (hunk 7 above)
Comment (Tal-or) [reply cmd/render/render.go:67]: Then we'll have to force the ```performance-profile.yaml``` to be on the same directory together with the other assets.
Is that something we want?

File: cmd/render/render.go (hunk 3 above)
Comment (cynepco3hahue) [reply cmd/render/render.go:31]: In general, I prefer `"k8s.io/klog"` because it used by k8s, also I can find only a few places where `log "github.com/sirupsen/logrus"` used.

File: cmd/render/render.go (hunk 9 above)
Comment (Tal-or) [reply cmd/render/render.go:114]: The names of the manifests are ```Kind.yaml``` so maybe instead of name call it ```fileName``` ?

File: cmd/render/render.go (hunk 8 above)
Comment (fromanirh) [reply cmd/render/render.go:69]: manifests are the assets PAO produces, and are assets in general

File: cmd/render/render.go (hunk 9 above)
Comment (fromanirh) [reply cmd/render/render.go:114]: let's keep like this for the time being and re-evaluate later.

File: cmd/render/render.go (hunk 10 above)
Comment (Tal-or) [reply cmd/render/render.go:79]: As for now we have no record for the possible objects that we're creating. it is possible to create a central place like pool/list/file which contains all the objects we have and keep it always updated, and both online/offline paths will follow it (no sure about the implementation details and maybe it deserve a separate PR).
But the same question can also asked here, what if forget to update this central place?

File: cmd/render/render.go (hunk 8 above)
Comment (Tal-or) [reply cmd/render/render.go]: Got it. Done

File: cmd/render/render.go (hunk 10 above)
Comment (fromanirh) [reply cmd/render/render.go:89]: indeed the problem is we have two completely separate flows :)

Comment (fromanirh) [reply cmd/render/render.go:89]: for the future:  I think we should evaluate a refactoring/code extraction from  controllers/performanceprofile_controller.go:applyComponents

File: vendor/modules.txt [hunk 12]
```diff
@@ -132,7 +132,8 @@ github.com/google/go-cmp/cmp/internal/function
 github.com/google/go-cmp/cmp/internal/value
```
Comment (fromanirh) [inline vendor/modules.txt:132]: Shouldn't this commit also include the updates to `go.mod` and `go.sum` ?

File: deploy/olm-catalog/performance-addon-operator/4.8.0/performance-addon-operator.v4.8.0.clusterserviceversion.yaml [hunk 13]
```diff
@@ -119,7 +119,8 @@ metadata:
       ]
     capabilities: Basic Install
```
Comment (fromanirh) [inline deploy/olm-catalog/performance-addon-operator/4.8.0/performance-addon-operator.v4.8.0.clusterserviceversion.yaml:120]: please ignore changes on clusterserviceversion file and merge changes to `go.mod` and `go.sum` in the previous commit.

File: main.go [hunk 14]
```diff
@@ -37,6 +38,7 @@ import (
 	// Import all Kubernetes client auth plugins (e.g. Azure, GCP, OIDC, etc.)
 	_ "k8s.io/client-go/plugin/pkg/client/auth"
 
+	render "github.com/openshift-kni/performance-addon-operators/pkg/cmd/render"
```
Comment (fromanirh) [inline main.go:41]: the rename is unneeded, but no big deal. Is this done automatically by the tooling?

File: pkg/cmd/render/render.go [hunk 15]
```diff
@@ -0,0 +1,196 @@
...
+	"k8s.io/klog"
+)
+
+type renderOpts struct {
+	ppInputFileDir string
+	assetsInDir    string
+	assetsOutDir   string
+}
+
+type manifestable map[string]v1.Object
```
Comment (fromanirh) [inline pkg/cmd/render/render.go:46]: better use `manifestTable` (camelCase)

File: pkg/cmd/render/render.go [hunk 16]
```diff
@@ -0,0 +1,196 @@
...
+
+//NewRenderCommand creates a render command.
+func NewRenderCommand() *cobra.Command {
+	renderOpts := renderOpts{}
+
+	cmd := &cobra.Command{
+		Use:   "render",
+		Short: "Render performance-addon-operator manifests",
+		Long: `Render performance-addon-operator manifests creates the following manifests:
```
Comment (fromanirh) [inline pkg/cmd/render/render.go:55]: This gets obsolete fast. Better to either autogenerate the list of output manifests or just be generic and tell "creates the output manifests" or so.

File: pkg/cmd/render/render.go [hunk 17]
```diff
@@ -0,0 +1,196 @@
...
+		},
+	}
+
+	renderOpts.AddFlags(cmd.Flags())
+
+	return cmd
+}
+
+func (r *renderOpts) AddFlags(fs *pflag.FlagSet) {
+	fs.StringVar(&r.ppInputFileDir, "performance-profile-dir-path", r.ppInputFileDir, "Input path for the Performance Profile manifest's directory.")
```
Comment (fromanirh) [inline pkg/cmd/render/render.go:78]: do we need a separate directory? How other operators do?

File: pkg/cmd/render/render.go [hunk 18]
```diff
@@ -0,0 +1,196 @@
...
+	fs.StringVar(&r.assetsInDir, "asset-input-dir", components.AssetsDir, "Input path for the assets directory.")
+	fs.StringVar(&r.assetsOutDir, "asset-output-dir", r.assetsOutDir, "Output path for the rendered manifests.")
+	// environmet variables has precedence over standard input
+	r.readFlagsFromEnv()
+}
+
+func (r *renderOpts) readFlagsFromEnv() {
+	if ppDir := os.Getenv("PP_DIR"); len(ppDir) > 0 {
```
Comment (fromanirh) [inline pkg/cmd/render/render.go:86]: too terse: let's be consistent with the long options. So, if the long options is
```
--something-like-this
```
then the env variable should be
```
SOMETHING_LIKE_THIS
```

Besides, is this something the other operators do?

File: pkg/cmd/render/render.go [hunk 19]
```diff
@@ -0,0 +1,196 @@
...
+	return nil
+}
+
+func (r *renderOpts) Run() error {
+	PPFiles := getPPList(r.ppInputFileDir)
+	if len(PPFiles) == 0 {
+		return fmt.Errorf("no valid performance-profiles found under: %v", r.ppInputFileDir)
+	}
+
+	for idx, pp := range PPFiles {
```
Comment (fromanirh) [inline pkg/cmd/render/render.go:117]: this looks similar but different wrt to `applyComponents`. We should try harder NOT to have two similar but distinct flows. Can we factor out the two flows in a common path?

File: pkg/cmd/render/render.go [hunk 20]
```diff
@@ -0,0 +1,196 @@
...
+
+// getPPList return a list of performance-profile manifests under dirPath
+func getPPList(dirPath string) []string {
+	files, err := ioutil.ReadDir(dirPath)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	// some regex to look for performance-profile manifests in directory's path
+	reg, err := regexp.Compile(`[P,p].*[-,_]*[P,p].*\.y[a]*ml`)
```
Comment (fromanirh) [inline pkg/cmd/render/render.go:183]: why is this needed?

File: pkg/cmd/render/render.go (hunk 18 above)
Comment (Tal-or) [reply pkg/cmd/render/render.go:86]: During our engineering meeting we agreed to add the env variables option.
I'll change to the naming convention you just suggested

File: pkg/cmd/render/render.go (hunk 19 above)
Comment (Tal-or) [reply pkg/cmd/render/render.go:117]: I thought we agreed to dedicate a separate PR for this, but I can work on that np.

File: pkg/cmd/render/render.go (hunk 20 above)
Comment (Tal-or) [reply pkg/cmd/render/render.go:183]: I don't know the performance profile name and there might be other files in the directory. so i'm looking for yaml's file that matches this regex

File: pkg/cmd/render/render.go (hunk 18 above)
Comment (fromanirh) [reply pkg/cmd/render/render.go:86]: Sure, did you check what other operators consume the env vars?

File: pkg/cmd/render/render.go (hunk 19 above)
Comment (fromanirh) [reply pkg/cmd/render/render.go:117]: I don't recall, it could very much be :) but if so that PR should come right after this one, we should not carry this tech debt for too long.

File: main.go (hunk 14 above)
Comment (Tal-or) [reply main.go:41]: Good question i have no idea. maybe it's a copy-paste mistake

File: pkg/cmd/render/render.go (hunk 20 above)
Comment (fromanirh) [reply pkg/cmd/render/render.go:183]: How others operator detect the input file? Do they have a separate command line flag?
Another option could be just decode all the YAMLs in the input directory, and look for a performance profile. I actually like this more than looking for specific names.

File: pkg/cmd/render/render.go (hunk 16 above)
Comment (fromanirh) [reply pkg/cmd/render/render.go:55]: I can think few things, but I believe is not worth (yet). I'll just go for the more generic phrasing.

File: pkg/cmd/render/render.go (hunk 17 above)
Comment (Tal-or) [reply pkg/cmd/render/render.go:78]: they provided separate directories if needed. see [here](https://github.com/openshift/installer/blob/master/data/data/bootstrap/files/usr/local/bin/bootkube.sh.template#L151)
Besides that, I don't think we need to force the user to have the performance-profile manifests together with other input-assets. This is more flexible don't you think?

Comment (fromanirh) [reply pkg/cmd/render/render.go:78]: Looking at the example you provided, I see
```
		/usr/bin/cluster-config-operator render \
		--cluster-infrastructure-input-file=/assets/manifests/cluster-infrastructure-02-config.yml \
                ....
```
It seems to me a bit cleaner to just add something like
```
fs.StringVar(&r.performanceProfileInputFile, "performance-profile-input-file", r.performanceProfileInputFile, "Input path for the Performance Profile manifest.")
```
This way we get immediatley what we want, and we don't need anymore extra detection steps, e.g. with regexes.

Comment (Tal-or) [reply pkg/cmd/render/render.go:78]: This was the original implementation, but we agreed to have the ability to render multiple pp files. 
Maybe I can provide a list of files` names,  instead of single file name or directory name

Comment (fromanirh) [reply pkg/cmd/render/render.go:78]: Right. Having the option accept more than a single file (or having the operator accept the same option multiple times) seems the best option.

File: pkg/cmd/render/render.go (hunk 18 above)
Comment (Tal-or) [reply pkg/cmd/render/render.go:86]: IIRC other operator don't. But we wanted this option so it will be more natural when running render inside a container

File: main.go (hunk 14 above)
Comment (fromanirh) [reply main.go:41]: ok, no big deal anyway.

File: pkg/cmd/render/render.go (hunk 18 above)
Comment (fromanirh) [reply pkg/cmd/render/render.go:86]: ok no big deal if the naming is consistent.

File: pkg/cmd/render/render.go (hunk 17 above)
Comment (Tal-or) [reply pkg/cmd/render/render.go:78]: How we can implement the same for env variable?
Is something like that:
```PERFORMANCE_PROFILE_INPUT_FILES=<file1's path>,<file2's path>,<file3's path>```
or ```PERFORMANCE_PROFILE_INPUT_FILES=<file1's path>:<file2's path>:<file3's path>```
is okay?

Comment (fromanirh) [reply pkg/cmd/render/render.go:78]: yep, fine. Comma separated is a bit more common (while colon ":" is more uncommon) so I'd go for that.

File: controllers/performanceprofile_controller.go [hunk 21]
```diff
@@ -18,16 +18,17 @@ package controllers
 
 import (
 	"context"
+	e "errors"
```
Comment (fromanirh) [inline controllers/performanceprofile_controller.go:21]: well, this seems a bit too much

File: controllers/performanceprofile_controller.go [hunk 22]
```diff
@@ -18,16 +18,17 @@ package controllers
...
+	"github.com/openshift-kni/performance-addon-operators/pkg/controller/performanceprofile/components/kubeletconfig"
+	"github.com/openshift-kni/performance-addon-operators/pkg/controller/performanceprofile/components/runtimeclass"
+	"github.com/openshift-kni/performance-addon-operators/pkg/controller/performanceprofile/components/tuned"
```
Comment (fromanirh) [inline controllers/performanceprofile_controller.go:24]: please move back these imports below, because it breaks proper grouping:

1. stdlib
2. third party
3. our own modules

File: controllers/performanceprofile_controller.go [hunk 23]
```diff
@@ -540,3 +542,34 @@ func hasMatchingLabels(performanceprofile *performancev2.PerformanceProfile, mcp
 	}
 	return true
 }
+
+// CompList a map that contains Kind string as key and component's instances as value
+type CompList map[string]interface{}
```
Comment (fromanirh) [inline controllers/performanceprofile_controller.go:547]: Well, I'd use a plain old struct like
```
type ManifestResultSet struct {
       MachineConfig *machineconfigv1.MachineConfig
       KubeletConfig *machineconfigv1.KubeletConfig
       Tuned         *tunedv1.Tuned
       RuntimeClass  *nodev1beta1.RuntimeClass
}
```
Not like we add new outputs very day (or even every major...)
This saves type assertions and looks a bit neater.

File: pkg/cmd/render/render.go [hunk 24]
```diff
@@ -0,0 +1,162 @@
...
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package render
+
+import (
+	"errors"
+	"fmt"
+	"github.com/openshift-kni/performance-addon-operators/controllers"
```
Comment (fromanirh) [inline pkg/cmd/render/render.go:21]: please move this in the group below

File: pkg/cmd/render/render.go [hunk 25]
```diff
@@ -0,0 +1,162 @@
...
+	performancev2 "github.com/openshift-kni/performance-addon-operators/api/v2"
+	"github.com/openshift-kni/performance-addon-operators/pkg/controller/performanceprofile/components"
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/klog"
+)
+
+type renderOpts struct {
+	ppInputFiles ppFiles
```
Comment (fromanirh) [inline pkg/cmd/render/render.go:37]: let's use `performanceProfileInputFiles` for consistency

File: pkg/cmd/render/render.go [hunk 26]
```diff
@@ -0,0 +1,162 @@
...
+	assetsInDir  string
+	assetsOutDir string
+}
+
+type ppFiles []string
```
Comment (fromanirh) [inline pkg/cmd/render/render.go:42]: let's use `performanceProfileFiles` for consistency

File: controllers/performanceprofile_controller.go (hunk 21 above)
Comment (Tal-or) [reply controllers/performanceprofile_controller.go:21]: ```errors``` and ```err``` are taken

File: pkg/cmd/render/render.go [hunk 27]
```diff
@@ -0,0 +1,162 @@
...
+			return err
+		}
+		or := []v1.OwnerReference{
+			{
+				Kind: profile.Kind,
+				Name: profile.Name,
+			},
+		}
+
+		for name, component := range components {
```
Comment (fromanirh) [inline pkg/cmd/render/render.go:146]: if we want to keep the generic range, I'd add a method like
 ```
func (ms ManifestResultSet) ToObjects() []v1.Object
```

**Review by fromanirh — COMMENTED** [review]
thanks for the updates, getting there

File: controllers/performanceprofile_controller.go (hunk 21 above)
Comment (fromanirh) [reply controllers/performanceprofile_controller.go:21]: keep this "errors"...

File: controllers/performanceprofile_controller.go [hunk 28]
```diff
@@ -18,16 +18,17 @@ package controllers
...
 
 	performancev2 "github.com/openshift-kni/performance-addon-operators/api/v2"
 	"github.com/openshift-kni/performance-addon-operators/pkg/controller/performanceprofile/components"
-	"github.com/openshift-kni/performance-addon-operators/pkg/controller/performanceprofile/components/kubeletconfig"
 	"github.com/openshift-kni/performance-addon-operators/pkg/controller/performanceprofile/components/machineconfig"
 	profileutil "github.com/openshift-kni/performance-addon-operators/pkg/controller/performanceprofile/components/profile"
-	"github.com/openshift-kni/performance-addon-operators/pkg/controller/performanceprofile/components/runtimeclass"
-	"github.com/openshift-kni/performance-addon-operators/pkg/controller/performanceprofile/components/tuned"
 	tunedv1 "github.com/openshift/cluster-node-tuning-operator/pkg/apis/tuned/v1"
 	mcov1 "github.com/openshift/machine-config-operator/pkg/apis/machineconfiguration.openshift.io/v1"
```
Comment (fromanirh) [inline controllers/performanceprofile_controller.go:34]: ...and rename the one below like `apierrors` or `k8serrors`

File: controllers/performanceprofile_controller.go (hunk 23 above)
Comment (Tal-or) [reply controllers/performanceprofile_controller.go:547]: But then we won't be able to iterate the struct, when most of the action are common and can be done using loops.

Comment (fromanirh) [reply controllers/performanceprofile_controller.go:547]: see below about `ToObjects`. Lacking generics, lacking better ideas (which admittledly I don't have now) we prefer quasi-duplicate, type-specific code to scattered type assertions.

Comment (Tal-or) [reply controllers/performanceprofile_controller.go:547]: yet you'll need a way to iterate the struct inside the ```ToObject``` in order to create this ```[]v1.Object``` but I see your point. type assertion might be confusing and less readable

Comment (fromanirh) [reply controllers/performanceprofile_controller.go:547]: In `ToObject` we should convert each field explicitely (no generic iteration). The idea is to keep at least all the conversion in a single place and let not them spread around in the codebase.

File: pkg/cmd/render/render.go [hunk 29]
```diff
@@ -0,0 +1,162 @@
...
+			component.(v1.Object).SetOwnerReferences(or)
+			b, err := yaml.Marshal(component)
```
Comment (Tal-or) [inline pkg/cmd/render/render.go:148]: @fromanirh  Any idea how to marshal here in a loop? converting to object isn't enough

Comment (fromanirh) [reply pkg/cmd/render/render.go:148]: I will have a closer look later

File: go.mod [hunk 30]
```diff
@@ -18,6 +18,7 @@ require (
 	github.com/go-openapi/strfmt v0.19.5 // indirect
 	github.com/go-openapi/swag v0.19.9 // indirect
 	github.com/go-openapi/validate v0.19.8 // indirect
+	github.com/google/uuid v1.2.0 // indirect
```
Comment (fromanirh) [inline go.mod:21]: these aren't really autogenerated changes, but rather deps bump/update

File: controllers/performanceprofile_controller.go [hunk 31]
```diff
@@ -360,52 +358,36 @@ func (r *PerformanceProfileReconciler) applyComponents(profile *performancev2.Pe
...
 
-	// get mutated machine config
-	mc, err := machineconfig.New(r.AssetsDir, profile)
+	components, err := manifestset.GetNewComponents(profile, &r.AssetsDir)
 	if err != nil {
 		return nil, err
 	}
-	if err := controllerutil.SetControllerReference(profile, mc, r.Scheme); err != nil {
-		return nil, err
+	for _, componentObj := range components.ToObject() {
```
Comment (fromanirh) [inline controllers/performanceprofile_controller.go:365]: nit: should be `ToObjects` (plural)

File: pkg/cmd/render/render.go [hunk 32]
```diff
@@ -0,0 +1,165 @@
...
+//NewRenderCommand creates a render command.
+func NewRenderCommand() *cobra.Command {
+	renderOpts := renderOpts{}
+
+	cmd := &cobra.Command{
+		Use:   "render",
+		Short: "Render performance-addon-operator manifests",
+		Run: func(cmd *cobra.Command, args []string) {
+
+			if err := renderOpts.Validate(); err != nil {
```
Comment (fromanirh) [inline pkg/cmd/render/render.go:73]: I don't think this (and Run) should not be a method of `renderOpts`, but we can surely live with it

File: pkg/cmd/render/render.go [hunk 33]
```diff
@@ -0,0 +1,165 @@
...
+			componentObj.SetOwnerReferences(or)
+		}
+
+		for name, manifest := range components.ToManifestTable() {
+			b, err := yaml.Marshal(manifest)
+			if err != nil {
+				return err
+			}
+
+			fileName := fmt.Sprintf("%s-%d.yaml", name, idx)
```
Comment (fromanirh) [inline pkg/cmd/render/render.go:156]: ok, this is the biggest and last point. I believe the output name should be like `$INPUT_NAME_tuned.yaml` `$INPUT_NAME_kubeletconfig.yaml` and so forth. IOW, use kind (`in_this_format`, `NotLikeThis`) and not a index. In case of ambiguity, we shoudl differentiate with object name if possible

**Review by fromanirh — COMMENTED** [review]
let's fix the output names and we're good to go.

**Review by fromanirh — COMMENTED** [review]
/lgtm
thanks for the updates
